palace/linalg/vector.cpp

```diff
@@ -0,0 +1,547 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "vector.hpp"
+
+#include <cstdint>
+#include <random>
+#include <mfem/general/forall.hpp>
+
+namespace palace
+{
+
+ComplexVector::ComplexVector(int n) : x(2 * n)
+{
+  xr.MakeRef(x, 0, n);
+  xi.MakeRef(x, n, n);
+}
+
+ComplexVector::ComplexVector(const ComplexVector &y) : x(2 * y.Size())
+{
+  xr.MakeRef(x, 0, y.Size());
+  xi.MakeRef(x, y.Size(), y.Size());
+  Set(y.Real(), y.Imag());
+}
+
+ComplexVector::ComplexVector(const Vector &yr, const Vector &yi) : x(2 * yr.Size())
+{
+  MFEM_VERIFY(yr.Size() == yi.Size(),
+              "Mismatch in dimension of real and imaginary matrix parts in ComplexVector!");
+  xr.MakeRef(x, 0, yr.Size());
+  xi.MakeRef(x, yr.Size(), yr.Size());
+  Set(yr, yi);
+}
+
+ComplexVector::ComplexVector(const std::complex<double> *py, int n) : x(2 * n)
+{
+  xr.MakeRef(x, 0, n);
+  xi.MakeRef(x, n, n);
+  Set(py, n);
+}
+
+void ComplexVector::SetSize(int n)
+{
+  x.SetSize(2 * n);
+  xr.MakeRef(x, 0, n);
+  xi.MakeRef(x, n, n);
+}
+
+ComplexVector &ComplexVector::operator=(const ComplexVector &y)
+{
+  Set(y.Real(), y.Imag());
```

Comment:
As discussed offline this is actually on purpose to avoid unintentional vector resizes or memory management. I've updated the comment in the header, but this is really just a utility function for `Set` rather than copy assignment.